src/bmark/run_solver.jl

```diff
@@ -1,127 +1,68 @@
 
 #### Keyword arguments
+* `logger::AbstractLogger`: logger used to show the statistics. Recommended `NullLogger`
+  for nothing or `ConsoleLogger`. (default: `NullLogger`).
+* `skipif::Function`: function to be applied to a problem and return whether to skip it
+  (default: `x->false`)
 * `prune`: do not include skipped problems in the final statistics (default: `true`)
-* any other keyword argument accepted by `run_problem()`
+* any other keyword argument to be passed to the solvers
 
 #### Return value
-* an `Array(AbstractExecutionStats, nprobs)` where `nprobs` is the number of problems
-  in `problems` minus the skipped ones if `prune` is true.
+* a `DataFrame` where each line is a problem, minus the skipped ones if `prune` is true.
 """
-function solve_problems(solver :: Function, problems :: Any; prune :: Bool=true, kwargs...)
-  display_header()
+function solve_problems(solver :: Function, problems :: Any;
+                        logger :: AbstractLogger=NullLogger(),
+                        skipif :: Function=x->false,
+                        colstats :: Array{Symbol,1} = [:name, :nvar, :ncon, :status],
+                        prune :: Bool=true, kwargs...)
   nprobs = length(problems)
   solverstr = split(string(solver), ".")[end]
 
   counter_fields = collect(fieldnames(Counters))
-  types = [String;  Symbol;    Float64;       Float64;   Int;    Float64; fill(Int, length(counter_fields))]
-  names = [ :name; :status; :objective; :elapsed_time; :iter; :dual_feas;                    counter_fields]
+  ncounters = length(counter_fields)
+  types = [Int; String;   Int;   Int;  Symbol;    Float64;       Float64;   Int;    Float64; fill(Int, ncounters)]
+  names = [:id;  :name; :nvar; :ncon; :status; :objective; :elapsed_time; :iter; :dual_feas;                    counter_fields]
   stats = DataFrame(types, names, 0)
 
-  k = 0
-  for problem in problems
+  with_logger(logger) do
+    # TODO: Improve logging after #74
+    @info join(string.(colstats), "  ")
+  end
+
+  for (id,problem) in enumerate(problems)
+    problem_info = [id; problem.meta.name; problem.meta.nvar; problem.meta.ncon]
+    if skipif(problem)
+      prune || push!(stats, [problem_info; :exception; Inf; Inf; 0; Inf; fill(0, ncounters)])
+      finalize(problem)
+      continue
+    end
     try
-      s = solve_problem(solver, problem; kwargs...)
-      push!(stats, [problem.meta.name; s.status; s.objective; s.elapsed_time; s.iter; s.dual_feas;
+      s = solver(problem; kwargs...)
+      push!(stats, [problem_info; s.status; s.objective; s.elapsed_time; s.iter; s.dual_feas;
                     [getfield(s.counters, f) for f in counter_fields]])
     catch e
-      isa(e, SkipException) || rethrow(e)
-      prune || push!(stats, [problem.meta.name; :exception; Inf; Inf; -1; Inf;
-                             fill(-1, length(counter_fields))])
+      push!(stats, [problem_info; :exception; Inf; Inf; 0; Inf;
+                    fill(0, ncounters)])
```

Comment:
Would it be possible to add `string(e)` somewhere?

Reply:
I could create a new column on the DataFrame just for extra messages, does that seem interesting?

Reply:
Yes, I think it'd be useful.